### PR TITLE
do not use batch updates with oracle

### PR DESCRIPTION
**Pull request by efonsell**

### CODE REVIEW

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java [hunk 1]
```diff
@@ -211,42 +211,53 @@ public PreparedStatement createPreparedStatement(Connection connection) throws S
...
-                  return false;
-                }
-                Entry<String, String> var = variables.next();
-                ps.setInt(1, id);
-                ps.setInt(2, actionId);
-                ps.setString(3, var.getKey());
-                ps.setString(4, var.getValue());
-                return true;
+    if (sqlVariants.useBatchUpdate()) {
+      final Iterator<Entry<String, String>> variables = changedStateVariables.entrySet().iterator();
```
Comment (jsyrjala) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java]: Refactor code for batch-update and non-batch-update to separate methods?

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java [hunk 2]
```diff
@@ -455,26 +466,44 @@ public OptimisticLockKey mapRow(ResultSet rs, int rowNum) throws SQLException {
...
-        int[] updateStatuses = jdbc.batchUpdate(updateInstanceForExecutionQuery()
-            + " where id = ? and modified = ? and executor_id is null", batchArgs);
-        Iterator<Integer> idIt = ids.iterator();
-        for (int status : updateStatuses) {
-          idIt.next();
-          if (status == 0) {
-            idIt.remove();
-            if (ids.isEmpty()) {
+        if (sqlVariants.useBatchUpdate()) {
+          List<Object[]> batchArgs = new ArrayList<>(instances.size());
```
Comment (jsyrjala) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java]: Refactor code for batch-update and non-batch-update to separate methods?

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java [hunk 3]
```diff
@@ -455,26 +466,44 @@ public OptimisticLockKey mapRow(ResultSet rs, int rowNum) throws SQLException {
...
+            int updated = jdbc.update(
+                updateInstanceForExecutionQuery() + " where id = ? and modified = ? and executor_id is null", instance.id,
+                instance.modified);
+            if (updated == 1) {
+              ids.add(instance.id);
+            } else {
+              raceConditionDetected = true;
+            }
+          }
+          if (raceConditionDetected && ids.isEmpty()) {
```
Comment (jsyrjala) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java]: What if raceConditionDetected and ids is not empty?

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java [hunk 4]
```diff
@@ -642,51 +670,50 @@ public String getWorkflowInstanceState(int workflowInstanceId) {
...
-        .setStateText(rs.getString("state_text"))
-        .setActions(new ArrayList<WorkflowInstanceAction>())
-        .setNextActivation(toDateTime(rs.getTimestamp("next_activation")))
-        .setRetries(rs.getInt("retries"))
-        .setCreated(toDateTime(rs.getTimestamp("created")))
-        .setModified(toDateTime(rs.getTimestamp("modified")))
-        .setStarted(toDateTime(rs.getTimestamp("started")))
-        .setExecutorGroup(rs.getString("executor_group"))
-        .build();
+      return new WorkflowInstance.Builder() //
```
Comment (jsyrjala) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java]: Why the empty // markers?

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java [hunk 5]
```diff
@@ -642,51 +670,50 @@ public String getWorkflowInstanceState(int workflowInstanceId) {
...
-        .setExecutorId(rs.getInt("executor_id"))
-        .setType(WorkflowActionType.valueOf(rs.getString("type")))
-        .setState(rs.getString("state"))
-        .setStateText(rs.getString("state_text"))
-        .setUpdatedStateVariables(actionState)
-        .setRetryNo(rs.getInt("retry_no"))
-        .setExecutionStart(toDateTime(rs.getTimestamp("execution_start")))
-        .setExecutionEnd(toDateTime(rs.getTimestamp("execution_end")))
-        .build();
+      Map<String, String> actionState = actionStates.containsKey(actionId) ? actionStates.get(actionId) : EMPTY_STATE_MAP;
```
Comment (jsyrjala) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java]: Why the empty // markers?

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java (hunk 3 above)
Comment (efonsell) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java]: Then we return the ids we got

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java (hunk 4 above)
Comment (efonsell) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java]: For eclipse formatting

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java (hunk 5 above)
Comment (efonsell) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java]: For eclipse formatting

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java (hunk 1 above)
Comment (efonsell) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java]: Sure

File: nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java (hunk 2 above)
Comment (efonsell) [inline nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/WorkflowInstanceDao.java]: Sure

**Pull request merged**